### ISSUE TITLE
Implement `VOLUME_CONDITION` capability

### ISSUE THREAD

**Issue by pkalever** [ceph/ceph-csi#2794]
# Describe the feature you'd like to have #

Healer Controller checks if a volume is still mounted and usable. To check whether a volume is usable, the CSI driver shall check if filesystem is corrupted, whether there are bad blocks, etc. in this RPC.

Instead of adding a new RPC, we can leverage the existing NodeGetVolumeStats RPC.

In the NodeGetVolumeStatsRequest, there are volume_id, volume_path, and staging_target_path. CSI driver can figure out whether a volume is still mounted based on these parameters.

A new message volume_condition will be added to NodeGetVolumeStatsResponse. A new Node capability `VOLUME_CONDITION` will be added to indicate whether a CSI driver has implemented this function. In a volume_condition message, there is a boolean parameter abnormal indicating whether the volume is normal or not and a message that describes the details of the volume condition.

# What new functionality do you want?

Provide a way to check the volume health status

# What is the value to the end-user? (why is it a priority?) #

The healer controller can leverage this to auto-fix the volume without user intervention.

# Additional context #

Read more at https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1432-volume-health-monitor#add-node-volume-health-function

**Comment by humblec** [comment]
There is already issue exist for this https://github.com/ceph/ceph-csi/issues/1356 .. lets use that.

**Comment by pkalever** [comment]
Here is what we can do to validate volume_condition with different volume mode persistent volumes

* FileSystem Volume Mode
1. if mounter is userspace check if its associated process is running, we can make this by reading MetaStashFile
2. if FS is writable, Attempt to write something on the target path

* Block Volume Mode
1. if mounter is userspace check if its associated process is running, we can make this by reading MetaStashFile
2. attempt read first few blocks from the device path


Any other ideas?

@ceph/ceph-csi-contributors ?

**Comment by pkalever** [comment]
@Madhu-1 and I had an opportunity to discuss on this and the idea was to just implement this with a simple stat for now. And additionally for userspace mounters check for `ps`

Any more thoughts?

**Comment by humblec** [comment]
@pkalever isnt this issue duplicate of https://github.com/ceph/ceph-csi/issues/1356 ?

**Comment by pkalever** [comment]
> @pkalever isnt this issue duplicate of #1356 ?

Nope, this is a subtask, the issue you pointed out intends to implement this for both controller and node, but this issue is only covers node cap. Thanks!

**Comment by humblec** [comment]
> > @pkalever isnt this issue duplicate of #1356 ?
> 
> Nope, this is a subtask, the issue you pointed out intends to implement this for both controller and node, but this issue is only covers node cap. Thanks!

SubTasks can be part of Parent Issue Tracker.. Will add to that any way.. Thanks .
